### PR TITLE
[react-redux] Give `connect` default types

### PR DESCRIPTION
**Pull request by tansongyang**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

If changing an existing definition:
- [ ] Provide a URL to documentation or source code which provides context for the suggested changes: <<url here>>
- [ ] Increase the version number in the header if appropriate.
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

### CODE REVIEW

File: types/react-redux/react-redux-tests.tsx
Comment (clayne11) [inline types/react-redux/react-redux-tests.tsx]: This is incorrect - you've removed a `{}` in the 3rd generic position and turned this into a 3-arity version vs the original 4-arity.

**Review by clayne11 — CHANGES_REQUESTED** [review]

**Comment by clayne11** [comment]
Could you give some details for the reason of this change?

File: types/react-redux/react-redux-tests.tsx
Comment (tansongyang) [reply types/react-redux/react-redux-tests.tsx]: Thanks for catching that.

**Comment by tansongyang** [comment]
@clayne11 This was simply a convenience update. Sometimes, when I write out types for `connect`, it would be convenient to not have to type out generic types that I don't need.

**Review by clayne11 — COMMENTED** [review]
I'm not a huge fan of defaulting the parameters. I'd rather wait for a potential update to the TS language that allows for declaring some types and inferring the others.

I'd like to see some discussion from another maintainer to see what their thoughts are.

**Comment by tansongyang** [comment]
Fair enough. Since this isn't a bug fix or anything like that, I'm cool with letting it die if other maintainers don't like it either. I am curious, though, what would be the reasons not to do this?

**Comment by clayne11** [comment]
I'm kinda waiting for a fix for https://github.com/Microsoft/TypeScript/issues/10571 to land. I'm not sure if that's going to happen anytime soon though.

**Review by clayne11 — APPROVED** [review]
I suppose we can merge this. No one else seems to be commenting. `{}` is a sane and safe default.

**Comment by NicholasBoll** [comment]
This change causes the wrong overrides to be picked up. For instance:

```ts
connect(
  null,
  mapDispatchToProps,
)(TestComponent)
```

This gets labeled as the signature that pulls in `mergeProps`. Also providing `mapStateToProps` and `mapDispatchToProps` gets picked up as needing mergeProps. Why are people overriding generics anyways? With so many generics to set it is easy to make a mistake. The point of generics is to infer from the call signature. You can have very strongly typed code with almost no generics outside function definitions.

```ts
function createType<T>(value: T): { type: T } {
  return { type: value }
}

const type1 = createType('foo') // { type: string }
const type2 = createType<string>('foo') // { type: string } ... but... why?
const type3 = createType('foo' as 'foo') // { type: 'foo' }
const type4 = createType<'foo'>('foo') // { type: 'foo' } ... leaky abstraction
```

Why do you want to have the overhead of making a mistake? I find casting is preferable to generic overrides. If you want types to narrow, extend from primitives:

```ts
function createType<T extends string | number | boolean | object>(value: T): { type: T } {
  return { type: value }
}

const type1 = createType('foo') // { type: 'foo' }
const type2 = createType<string>('foo') // { type: string }
const type3 = createType('foo' as string) // { type: string }
const type4 = createType(true) // { type: true }
const type5 = createType(true as boolean) // { type: boolean }
const type6 = createType(1) // { type: 1 }
const type7 = createType(1 as number) // { type: number }
```

I can get away with not providing generics to `connect` except for 1 bug: `() => MappedState | () => () => MappedState` defines the factory second and thus has no call signatures. Simply reversing the order allows TypeScript to pick it up just fine.

**Comment by clayne11** [comment]
Alright, maybe we should revert this PR. I wasn't super sold on it but none of the other maintainers jumped in with another opinion and it seemed safe enough.

I agree with you @NicholasBoll, generally types should be inferred rather than set explicitly.

**Comment by tansongyang** [comment]
@NicholasBoll @clayne11 Thanks for checking that out. Hadn't thought of that. I think that's reason enough to revert. Working on it now.

**Comment by tansongyang** [comment]
PR that reverts this PR is [here](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/19651). Let's continue the discussion there.

**Pull request merged**
